ddmark/validation.go

```diff
@@ -0,0 +1,220 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+package ddmark
+
+import (
+	"fmt"
+	"reflect"
+
+	k8smarkers "sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+// AllDefinitions contains all marker definitions for this package.
+var AllDefinitions []*k8smarkers.Definition
+var rulePrefix string = "ddmark:validation:"
+
+func init() {
+	addDefinition(Maximum(0), k8smarkers.DescribesField)
+	addDefinition(Minimum(0), k8smarkers.DescribesField)
+	addDefinition(Enum(nil), k8smarkers.DescribesField)
+	addDefinition(Required(true), k8smarkers.DescribesField)
+
+	addDefinition(ExclusiveFields(nil), k8smarkers.DescribesType)
+	addDefinition(LinkedFields(nil), k8smarkers.DescribesType)
+}
+
+// Maximum can applied to an int field and provides a (non-strict) maximum value for that field
+type Maximum int
+
+// Minimum can applied to an int field and provides a (non-strict) minimum value for that field
+type Minimum int
+
+// Enum can be applied to any interface field and provides a restricted amount of possible values for that field.
+// Values within the marker strictly need to fit the given field interface. Usage is recommended to simple types.
+type Enum []interface{}
+
+// Required can be applied to any field, and asserts this field will return an error if not provided
+type Required bool
+
+// ExclusiveFields can be applied to structs, and asserts that at most one of the specified fields is set
+type ExclusiveFields []string
+
+// LinkedFields can be applied to structs, and asserts that if the first field isn't nil, then non of the others are either
```

Comment:
```suggestion
// LinkedFields can be applied to structs, and asserts that if the first field isn't nil, then none of the others are either
```

could even be "all of them are set, or none of them are"